### PR TITLE
feat: Persistence cheatcodes

### PR DESCRIPTION
**Pull request by Jrigada**
# What :computer: 
* Marking, revoking and checking persistent accounts across forks. 

# Why :hand:

# Evidence :camera:
image

image

### CODE REVIEW

File: crates/era-cheatcodes/tests/src/cheatcodes/Fork.t.sol [hunk 1]
```diff
@@ -78,4 +79,44 @@ contract ForkTest is Test {
...
+        require(vm.isPersistent(address(this)) == true, "should be persistent");
+
+        // the dummy address should not be persistent
+        require(
+            vm.isPersistent(DUMMY_ADDRESS) == false,
+            "should not be persistent"
+        );
+
+        // mark the dummy address as persistent
+        vm.makePersistent(DUMMY_ADDRESS);
```
Comment (Deniallugo) [inline crates/era-cheatcodes/tests/src/cheatcodes/Fork.t.sol:94]: Could we check it together with `selectFork` that we are actually doing everything correct

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 2]
```diff
@@ -1143,6 +1195,18 @@ impl CheatcodeTracer {
...
+                tracing::info!("👷 Checking if account {:?} is persistent", account);
+                let handle: &ForkStorage<RevmDatabaseForEra<S>> =
+                    &storage.borrow_mut().storage_handle;
+                let fork_storage = handle.inner.read().unwrap();
+                let era_db = fork_storage.fork.as_ref().unwrap().fork_source.clone();
+                let db = era_db.db.lock().unwrap();
+                let is_persistent = db.is_persistent(&revm::primitives::Address::from(
+                    account.to_h160().to_fixed_bytes(),
+                ));
+                self.return_data = Some(vec![U256::from(is_persistent as u8)]);
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:1208]:
```suggestion
                self.return_data = Some(is_persistent.to_return_data());
```

File: crates/zkforge/src/runner.rs [hunk 3]
```diff
@@ -201,6 +203,51 @@ impl<'a> ContractRunner<'a> {
         Ok(setup)
     }
 
+    /// Sets up the cheatcode address with an empty contract so `vm.func()` calls do not panic.
+    fn setup_cheatcode_address(&mut self) {
```
Comment (nbaztec) [inline crates/zkforge/src/runner.rs:207]: This function is removed

File: crates/zkforge/src/runner.rs [hunk 4]
```diff
@@ -212,6 +259,8 @@ impl<'a> ContractRunner<'a> {
         let start = Instant::now();
         let mut warnings = Vec::new();
 
+        self.setup_cheatcode_address();
```
Comment (nbaztec) [inline crates/zkforge/src/runner.rs:262]: Also this is removed

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 5]
```diff
@@ -640,6 +644,54 @@ impl<S: DatabaseExt + Send, H: HistoryMode> VmTracer<EraDb<S>, H> for CheatcodeT
...
+                    let mut db = era_db.db.lock().unwrap();
+                    db.extend_persistent_accounts(
+                        accounts
+                            .into_iter()
+                            .map(|a: H160| revm::primitives::Address::from(a.to_fixed_bytes()))
+                            .collect::<Vec<revm::primitives::Address>>(),
+                    );
+                }
+                FinishCycleOneTimeActions::RevokePersistentAccount { account } => {
+                    let handle: &ForkStorage<RevmDatabaseForEra<S>> =
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:673]: Are these type annotations required here?

Comment (Jrigada) [reply crates/era-cheatcodes/src/cheatcodes.rs:673]: Not really, I'll remove it from the one time actions.

File: crates/era-cheatcodes/tests/src/cheatcodes/Persistent.t.sol [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+contract ForkPersistentTest is Test {
+    uint256 constant FORK_BLOCK = 19579636;
+    address constant DUMMY_ADDRESS = 0x1345df6d4C9c3035724fd0e3914dE96a5a83AAF4;
+
+    /// checks that marking as persistent works
+    function testMakePersistent() public {
+        uint256 mainnetFork = vm.createFork("mainnet", FORK_BLOCK + 100);
+        uint256 optimismFork = vm.createFork("mainnet", FORK_BLOCK + 100);
```
Comment (Deniallugo) [inline crates/era-cheatcodes/tests/src/cheatcodes/Persistent.t.sol:32]: so, we have two identical forks, is it on purpose? and why do you call one of them optimism?

Comment (Jrigada) [reply crates/era-cheatcodes/tests/src/cheatcodes/Persistent.t.sol:32]: Maybe the names are misleading, I'll change them. It is similar to how foundry test MakePersistent.

**Comment by nbaztec** [comment]
@Jrigada please rebase on main, then this can be merged

File: crates/era-cheatcodes/src/cheatcodes.rs [hunk 7]
```diff
@@ -806,6 +849,9 @@ impl<S: DatabaseExt + Send, H: HistoryMode> VmTracer<EraDb<S>, H> for CheatcodeT
                     let mut storage = storage.borrow_mut();
                     let modified_storage =
                         self.get_modified_storage(storage.modified_storage_keys());
+                    let modified_bytecodes = self.get_modified_bytecodes(
+                        bootloader_state.get_last_tx_compressed_bytecodes(),
+                    );
```
Comment (nbaztec) [inline crates/era-cheatcodes/src/cheatcodes.rs:854]: same as above

**Review by Deniallugo — APPROVED** [review]

**Pull request merged**